File: Server~/src/tools/updateGameObjectTool.ts
import * as z from 'zod';
import { McpUnity } from '../unity/mcpUnity.js';
import { Logger } from '../utils/logger.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';
import { payloadContent } from '../utils/toolPayload.js';

// Constants for the tool
const toolName = 'update_gameobject';
const toolDescription = 'Updates properties of a GameObject in the Unity scene by its instance ID or path. If the GameObject does not exist at the specified path, it will be created. Every supplied gameObjectData key is reported in updatedFields or failedFields; valid fields may still be applied when another field fails, and any field failure returns isError=true with the full payload.';
const paramsSchema = z.object({
  instanceId: z.number().optional().describe('The instance ID of the GameObject to update'),
  objectPath: z.string().optional().describe('The path of the GameObject in the hierarchy to update (alternative to instanceId)'),
  gameObjectData: z.object({
    name: z.string().optional().describe('New name for the GameObject'),
    tag: z.string().optional().describe('New tag for the GameObject'),
    layer: z.number().int().optional().describe('New layer for the GameObject'),
    activeSelf: z.boolean().optional().describe('Set the active state of the GameObject (GameObject.SetActive(value))'),
    isStatic: z.boolean().optional().describe('Set the static state of the GameObject (GameObject.isStatic = value)'),
  }).strict().describe('An object containing supported fields to update on the GameObject. Unknown keys are rejected instead of being ignored. If the GameObject does not exist at objectPath, it will be created.')
    .refine(data => Object.keys(data).length > 0, { message: 'gameObjectData must contain at least one property to update.' }),
});

/**
 * Creates and registers the Update GameObject tool with the MCP server
 * This tool allows updating or creating GameObjects into the Unity Editor active scene
 * 
 * @param server The MCP server instance to register with
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param logger The logger instance for diagnostic information
 */
export function registerUpdateGameObjectTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
    logger.info(`Registering tool: ${toolName}`);
        
    // Register this tool with the MCP server
    server.tool(
      toolName,
      toolDescription,
      paramsSchema.shape,
      async (params: any) => {
        try {
          logger.info(`Executing tool: ${toolName}`, params);
          const result = await toolHandler(mcpUnity, params);
          logger.info(`Tool execution successful: ${toolName}`);
          return result;
        } catch (error) {
          logger.error(`Tool execution failed: ${toolName}`, error);
          throw error;
        }
      }
    );
}
/**
 * Handles updating or creating GameObjects into the Unity Editor active scene
 * 
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param params The parameters for the tool
 * @returns A promise that resolves to the tool execution result
 * @throws McpUnityError if the request to Unity fails
 */
async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
    // Validate parameters - require either instanceId or objectPath
    if ((params.instanceId === undefined || params.instanceId === null) && 
        (!params.objectPath || params.objectPath.trim() === '')) {
        throw new McpUnityError(
        ErrorType.VALIDATION,
        "Either 'instanceId' or 'objectPath' must be provided"
        );
    }

    const response = await mcpUnity.sendRequest({
        method: 'update_gameobject',
        params: {
          instanceId: params.instanceId,
          objectPath: params.objectPath,
          gameObjectData: params.gameObjectData,
        }
      });
  
      const hasFieldFailures = Array.isArray(response.failedFields) && response.failedFields.length > 0;

      if (!response.success && !hasFieldFailures) {
        throw new McpUnityError(
          ErrorType.TOOL_EXECUTION,
          response.message || `Failed to update the GameObject`
        );
      }
  
      // Create a description of which GameObject was targeted
      const targetDescription = params.objectPath 
        ? `path '${params.objectPath}'` 
        : `ID ${params.instanceId}`;
      
      const result: CallToolResult = {
        content: [
          {
            type: response.type || 'text',
            text: response.message || `Successfully updated the GameObject with ${targetDescription}`
          },
          payloadContent({
            message: response.message,
            instanceId: response.instanceId,
            name: response.name,
            path: response.path,
            updatedFields: response.updatedFields,
            failedFields: response.failedFields,
            warnings: response.warnings
          })
        ]
      };

      if (hasFieldFailures) {
        result.isError = true;
      }

      return result;
}

File: Editor/Tools/UpdateGameObjectTool.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.UI;
using UnityEditor;
using McpUnity.Utils; // For GameObjectHierarchyCreator and McpLogger
using McpUnity.Unity; // For McpUnitySocketHandler
using McpUnity.Services; // For PrefabEditingService
using Newtonsoft.Json.Linq; // For JObject

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for updating or creating a GameObject in the Unity Editor.
    /// Supports setting name, tag, layer, active state, and static state by instance ID or hierarchy path.
    /// Returns a JObject result similar to UpdateComponentTool for consistency.
    /// </summary>
    public class UpdateGameObjectTool : McpToolBase
    {
        private static readonly HashSet<string> SupportedGameObjectFields =
            new HashSet<string>(StringComparer.Ordinal)
            {
                "name",
                "tag",
                "layer",
                "activeSelf",
                "isActiveSelf",
                "isStatic",
                "static"
            };

        public UpdateGameObjectTool()
        {
            Name = "update_gameobject";
            Description = "Updates or creates a GameObject and its properties (name, tag, layer, active state, static state) based on instance ID or object path. Every supplied gameObjectData key is reported in updatedFields or failedFields; valid fields may still be applied when another field fails.";
            IsAsync = false; // Operations are expected to be quick
        }

        /// <summary>
        /// Executes the update or creation of a GameObject based on the provided parameters.
        /// </summary>
        /// <param name="parameters">A JObject containing: instanceId (int?), objectPath (string), name (string), tag (string), layer (int?), isActiveSelf (bool?), isStatic (bool?)</param>
        /// <returns>JObject with success, message, instanceId, name, and path fields (see UpdateComponentTool for format)</returns>
        public override JObject Execute(JObject parameters)
        {
            // Extract parameters from JObject
            int? instanceId = parameters["instanceId"]?.ToObject<int?>();
            string objectPath = parameters["objectPath"]?.ToObject<string>();
            JObject gameObjectData = parameters["gameObjectData"] as JObject;

            GameObject targetGameObject = null;
            string identifierInfo = "";

            // Identify or create the GameObject by instanceId or objectPath
            if (instanceId.HasValue)
            {
                JObject scopeError = PrefabSessionScope.TryResolveGameObject(
                    instanceId, null, out targetGameObject);
                if (scopeError != null) return scopeError;
                identifierInfo = $"instance ID {instanceId.Value}";
            }
            else if (!string.IsNullOrEmpty(objectPath))
            {
                JObject createError = GameObjectHierarchyCreator.TryFindOrCreateHierarchicalGameObject(
                    objectPath, out targetGameObject);
                if (createError != null) return createError;
                identifierInfo = $"path '{objectPath}'";
            }
            else
            {
                // Neither instanceId nor objectPath was provided
                return McpUnitySocketHandler.CreateErrorResponse("Either 'instanceId' or 'objectPath' must be provided.", "validation_error");
            }

            // Check if we could not identify or create the GameObject
            if (targetGameObject == null)
            {
                return McpUnitySocketHandler.CreateErrorResponse($"Target GameObject could not be identified or created using {identifierInfo}.", "unknown_error");
            }

            // Record for undo in Unity Editor
            Undo.RecordObject(targetGameObject, "Update GameObject Properties");
            bool propertiesUpdated = false;
            string originalNameForLog = targetGameObject.name;
            var updatedFields = new List<string>();
            var failedFields = new List<JObject>();
            var warnings = new List<string>();

            JProperty nameProperty = gameObjectData?.Property("name");
            if (nameProperty != null)
            {
                if (nameProperty.Value.Type != JTokenType.String
                    || string.IsNullOrEmpty(nameProperty.Value.ToObject<string>()))
                {
                    failedFields.Add(CreateFieldFailure(
                        "name", "Name must be a non-empty JSON string. Name was not changed."));
                }
                else
                {
                    string newName = nameProperty.Value.ToObject<string>();
                    if (targetGameObject.name != newName)
                    {
                        targetGameObject.name = newName;
                        propertiesUpdated = true;
                    }
                    updatedFields.Add("name");
                }
            }

            JProperty tagProperty = gameObjectData?.Property("tag");
            if (tagProperty != null)
            {
                string newTag = tagProperty.Value.Type == JTokenType.String
                    ? tagProperty.Value.ToObject<string>()
                    : null;
                bool tagExists = !string.IsNullOrEmpty(newTag)
                    && Array.Exists(
                        UnityEditorInternal.InternalEditorUtility.tags,
                        tag => tag.Equals(newTag));
                if (!tagExists)
                {
                    string reason = $"Tag '{newTag}' does not exist. Tag was not changed; " +
                        "create it in Unity's Tag Manager before retrying.";
                    failedFields.Add(CreateFieldFailure("tag", reason));
                    McpLogger.LogWarning(
                        $"UpdateGameObjectTool: {reason} GameObject: '{originalNameForLog}'.");
                }
                else
                {
                    if (!targetGameObject.CompareTag(newTag))
                    {
                        targetGameObject.tag = newTag;
                        propertiesUpdated = true;
                    }
                    updatedFields.Add("tag");
                }
            }

            JProperty layerProperty = gameObjectData?.Property("layer");
            if (layerProperty != null)
            {
                if (layerProperty.Value.Type != JTokenType.Integer)
                {
                    failedFields.Add(CreateFieldFailure(
                        "layer", "Layer must be a JSON integer between 0 and 31. Layer was not changed."));
                }
                else
                {
                    long newLayer = layerProperty.Value.ToObject<long>();
                    if (newLayer < 0 || newLayer > 31)
                    {
                        string reason = $"Layer value {newLayer} is outside the valid range 0-31. " +
                            "Layer was not changed.";
                        failedFields.Add(CreateFieldFailure("layer", reason));
                        McpLogger.LogWarning(
                            $"UpdateGameObjectTool: {reason} GameObject: '{originalNameForLog}'.");
                    }
                    else
                    {
                        if (targetGameObject.layer != (int)newLayer)
                        {
                            targetGameObject.layer = (int)newLayer;
                            propertiesUpdated = true;
                        }
                        updatedFields.Add("layer");
                    }
                }
            }

            JProperty activeSelfProperty = gameObjectData?.Property("activeSelf");
            JProperty legacyActiveSelfProperty = gameObjectData?.Property("isActiveSelf");
            propertiesUpdated |= ApplyBooleanField(
                activeSelfProperty,
                value => targetGameObject.activeSelf != value,
                value => targetGameObject.SetActive(value),
                updatedFields,
                failedFields);
            if (legacyActiveSelfProperty != null && activeSelfProperty != null)
            {
                failedFields.Add(CreateFieldFailure(
                    "isActiveSelf",
                    "Both 'activeSelf' and legacy alias 'isActiveSelf' were provided; use only 'activeSelf'."));
            }
            else
            {
                propertiesUpdated |= ApplyBooleanField(
                    legacyActiveSelfProperty,
                    value => targetGameObject.activeSelf != value,
                    value => targetGameObject.SetActive(value),
                    updatedFields,
                    failedFields);
            }

            JProperty isStaticProperty = gameObjectData?.Property("isStatic");
            JProperty legacyStaticProperty = gameObjectData?.Property("static");
            propertiesUpdated |= ApplyBooleanField(
                isStaticProperty,
                value => targetGameObject.isStatic != value,
                value => targetGameObject.isStatic = value,
                updatedFields,
                failedFields);
            if (legacyStaticProperty != null && isStaticProperty != null)
            {
                failedFields.Add(CreateFieldFailure(
                    "static",
                    "Both 'isStatic' and legacy alias 'static' were provided; use only 'isStatic'."));
            }
            else
            {
                propertiesUpdated |= ApplyBooleanField(
                    legacyStaticProperty,
                    value => targetGameObject.isStatic != value,
                    value => targetGameObject.isStatic = value,
                    updatedFields,
                    failedFields);
            }

            if (gameObjectData != null)
            {
                foreach (JProperty property in gameObjectData.Properties())
                {
                    if (!SupportedGameObjectFields.Contains(property.Name))
                    {
                        failedFields.Add(CreateFieldFailure(
                            property.Name,
                            $"Unknown GameObject field '{property.Name}'. Valid fields: " +
                            "name, tag, layer, activeSelf, isActiveSelf, isStatic, static."));
                    }
                }
            }

            // Mark as dirty if any property was changed
            if (propertiesUpdated)
            {
                EditorUtility.SetDirty(targetGameObject);
            }

            // Check if the GameObject is under a Canvas but lacks RectTransform
            if (targetGameObject.GetComponentInParent<Canvas>() != null
                && targetGameObject.GetComponent<RectTransform>() == null)
            {
                warnings.Add("This GameObject is under a Canvas but has no RectTransform. "
                    + "Use 'create_ui_element' for UI objects to get RectTransform automatically, "
                    + "or add a RectTransform via 'update_component'.");
            }

            // Compose result message and return as JObject (like UpdateComponentTool)
            string message = $"Updated GameObject '{targetGameObject.name}' (identified by {identifierInfo}): " +
                $"{updatedFields.Count} field(s) succeeded, {failedFields.Count} field(s) failed";
            if (warnings.Count > 0)
            {
                message += $" (with {warnings.Count} warning(s))";
            }

            return new JObject
            {
                ["success"] = failedFields.Count == 0,
                ["type"] = "text",
                ["message"] = message,
                ["instanceId"] = targetGameObject.GetInstanceID(),
                ["name"] = targetGameObject.name,
                ["path"] = GameObjectPathUtils.GetCanonicalPath(targetGameObject),
                ["updatedFields"] = new JArray(updatedFields.ToArray()),
                ["failedFields"] = new JArray(failedFields.ToArray()),
                ["warnings"] = new JArray(warnings.ToArray())
            };
        }

        private static JObject CreateFieldFailure(string fieldName, string reason)
        {
            return new JObject
            {
                ["field"] = fieldName,
                ["reason"] = reason
            };
        }

        private static bool ApplyBooleanField(
            JProperty property,
            Func<bool, bool> differsFromCurrent,
            Action<bool> apply,
            List<string> updatedFields,
            List<JObject> failedFields)
        {
            if (property == null)
                return false;

            if (property.Value.Type != JTokenType.Boolean)
            {
                failedFields.Add(CreateFieldFailure(
                    property.Name,
                    $"Field '{property.Name}' must be a JSON boolean. Field was not changed."));
                return false;
            }

            bool value = property.Value.ToObject<bool>();
            bool changed = differsFromCurrent(value);
            if (changed)
            {
                apply(value);
            }
            updatedFields.Add(property.Name);
            return changed;
        }

    }
}
